### PR TITLE
[full-ci] Store settings via metadata

### PR DESCRIPTION
**Pull request by kobergj**
## Description

Encourages settings service to store its data via the metadata service

## Related Issue

- Fixes https://jira.owncloud.com/browse/OCIS-2346

## How Has This Been Tested?

- new unit tests
- existing acceptance tests cover functionality

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>Core-API-Tests-ocis-storage-5</strong>](https://drone.owncloud.com/owncloud/ocis/9663/26/1) failed. Further test are cancelled...

File: settings/pkg/service/v0/service.go [hunk 1]
```diff
@@ -32,12 +33,21 @@ type Service struct {
...
+
+	switch cfg.StoreType {
+	default:
+		fallthrough
+	case "metadata":
+		service.manager = metastore.New(cfg)
+	case "filesystem":
+		service.manager = filestore.New(cfg)
+		// TODO: if we want to further support filesystem store it should use default permissions from store/defaults/defaults.go instead using this duplicate
+		service.RegisterDefaultRoles()
```
Comment (wkloucek) [inline settings/pkg/service/v0/service.go:49]: I don't see any roles stored in the metadata service. Shouldn't we call RegisterDefaultRoles in all cases?

**Review by wkloucek — CHANGES_REQUESTED** [review]
I couldn't get it working. Eg. as admin I am not allowed to add / remove users:
image

**Comment by kobergj** [comment]
@wkloucek Creating users works fine for me. Currently it is necessary to set `OCIS_URL` envvar because of the IDP of the service user. Could you try if this is the issue?

File: settings/pkg/service/v0/service.go (hunk 1 above)
Comment (kobergj) [reply settings/pkg/service/v0/service.go:49]: No. Metadata store package is setting the default roles form `store/defaults/defaults.go` This is a duplicate.

**Comment by C0rby** [comment]
Yeah, it doesn't work on my machine either. I can login but I can't access the settings/accounts UI.
I get the following errors in my log:
```
2022-03-14T11:50:36+01:00 ERR error initializing metadata client error="error: permission denied: permission denied" service=ocis
2022-03-14T11:50:45+01:00 WRN Failed to execute error="{\"id\":\"com.owncloud.api.thumbnails\",\"code\":404,\"detail\":\"Unsupported file type\",\"status\":\"Not Found\"}" duration=3.591365 method=Thumbnails.GetThumbnail service=thumbnails
2022-03-14T11:51:04+01:00 ERR could not read link, skipping error="readlink /home/corby/.ocis/storage/users/spacetypes/personal/ddc2004c-0977-11eb-9d3f-a793888cd0f8.flock: no such file or directory" match=/home/corby/.ocis/storage/users/spacetypes/personal/ddc2004c-0977-11eb-9d3f-a793888cd0f8.flock pkg=rgrpc service=storage traceid=00000000000000000000000000000000
2022-03-14T11:51:16+01:00 ERR could not read link, skipping error="readlink /home/corby/.ocis/storage/users/spacetypes/personal/ddc2004c-0977-11eb-9d3f-a793888cd0f8.flock: no such file or directory" match=/home/corby/.ocis/storage/users/spacetypes/personal/ddc2004c-0977-11eb-9d3f-a793888cd0f8.flock pkg=rgrpc service=storage traceid=00000000000000000000000000000000
2022-03-14T11:52:15+01:00 ERR error initializing metadata client error="error: not found: stat: error: not found: f1bdd61a-da7c-49fc-8203-0558109d1b4f/settings" service=ocis
2022-03-14T11:52:21+01:00 ERR error initializing metadata client error="error: not found: stat: error: not found: f1bdd61a-da7c-49fc-8203-0558109d1b4f/settings" service=ocis
2022-03-14T11:52:29+01:00 ERR error initializing metadata client error="error: not found: stat: error: not found: f1bdd61a-da7c-49fc-8203-0558109d1b4f/settings" service=ocis
```

Started oCIS like this: `OCIS_URL=https://localhost:9200 OCIS_INSECURE=true PROXY_ENABLE_BASIC_AUTH=true OCIS_LOG_PRETTY=true OCIS_LOG_COLOR=t
rue OCIS_LOG_LEVEL=warn bin/ocis server`

**Comment by kobergj** [comment]
@C0rby this is happening when the settings folder could not be created properly. Could you remove `~/.ocis/storage/metadata/spaces/f1` and try again?

**Comment by C0rby** [comment]
Starting with a clean storage and the `OCIS_URL` works.

**Comment by kobergj** [comment]
Ok cool thanks :+1: Imho it is acceptable to have to remove `metadata/spaces/f1` if it already exists (it shouldn't normally)

Should we do something about the `OCIS_URL`? I can hardcode the idp instead of taking it from env...

File: settings/pkg/config/defaults/defaultconfig.go [hunk 2]
```diff
@@ -44,13 +46,22 @@ func DefaultConfig() *config.Config {
...
 		},
 		TokenManager: config.TokenManager{
 			JWTSecret: "Pive-Fumkiu4",
 		},
+
+		Metadata: config.Metadata{
+			GatewayAddress:    "127.0.0.1:9142",
+			StorageAddress:    "127.0.0.1:9215",
+			ServiceUserID:     "ddc2004c-0977-11eb-9d3f-a793888cd0f8",
+			ServiceUserIDP:    os.Getenv("OCIS_URL"),
```
Comment (wkloucek) [inline settings/pkg/config/defaults/defaultconfig.go:62]: I really dislike the service user idp setting, but as far as I understood we cannot easily work around it. I think we should at least configure it like the other OCIS_URL / IDP URL settings. (see eghttps://github.com/owncloud/ocis/blob/51d1eb0182fce0ed3920689ac3ed5383b2eb2e84/idp/pkg/config/config.go#L56, https://github.com/owncloud/ocis/blob/51d1eb0182fce0ed3920689ac3ed5383b2eb2e84/idp/pkg/config/defaults/defaultconfig.go#L38)

File: settings/pkg/config/defaults/defaultconfig.go [hunk 3]
```diff
@@ -44,13 +46,22 @@ func DefaultConfig() *config.Config {
...
 			Path: "",
 		},
 		TokenManager: config.TokenManager{
 			JWTSecret: "Pive-Fumkiu4",
 		},
+
+		Metadata: config.Metadata{
+			GatewayAddress:    "127.0.0.1:9142",
+			StorageAddress:    "127.0.0.1:9215",
+			ServiceUserID:     "ddc2004c-0977-11eb-9d3f-a793888cd0f8",
```
Comment (wkloucek) [inline settings/pkg/config/defaults/defaultconfig.go:61]: should we really use the admin user here? why can't we use a system user like for the accounts service?

File: settings/pkg/config/defaults/defaultconfig.go (hunk 2 above)
Comment (kobergj) [reply settings/pkg/config/defaults/defaultconfig.go]: :+1: That's what I meant with "hardcoding it"

File: settings/pkg/config/defaults/defaultconfig.go (hunk 3 above)
Comment (kobergj) [reply settings/pkg/config/defaults/defaultconfig.go]: Well. Seems like it still doesn't work :disappointed:

Comment (kobergj) [reply settings/pkg/config/defaults/defaultconfig.go]: @wkloucek I found a way around it. Can you check again?

**Review by wkloucek — APPROVED** [review]

**Pull request merged**